lib/awful/keygrabber.lua

```diff
@@ -45,7 +207,411 @@ function keygrabber.stop(g)
+-- @tparam table mod The current modifiers (like "Control" or "Shift").
+-- @tparam string key The key name.
+-- @tparam string event The event ("press" or "release")
+
+--- A wrapper around the keygrabber to make it easier to add keybindings.
+--
+-- This is similar to the `awful.prompt`, but without an actual widget. It can
+-- be used to implement custom transactions such as alt+tab or keyboard menu
+-- navigation.
+--
+-- Note that the object returned can be used as a client or global keybinding
+-- callback without modification. Make sure to set `release_key` and `release_event`
+-- to proper values to avoid permanently locking the keyboard.
+--
+-- @tparam table args
+-- @tparam[opt="press"] string args.release_event Release event ("press" or "release")
+-- @tparam[opt=nil] string|table args.release_key The key that has to be kept pressed.
+-- @tparam table args.keybindings All keybindings.
+-- @tparam[opt=-1] number args.timeout The maximum inactivity delay.
+-- @tparam[opt=true] boolean args.mask_event_callback Do not call `keypressed_callback`
+--  or `keyreleased_callback` when a hook is called.
+-- @tparam[opt] function args.start_callback
+-- @tparam[opt] function args.stop_callback
+-- @tparam[opt] function args.timeout_callback
+-- @tparam[opt] function args.keypressed_callback
+-- @tparam[opt] function args.keyreleased_callback
+-- @tparam[opt=nil] table|nil args.allowed_keys A table with all keys that
+--  **wont** stop the keygrabber.
+-- @tparam[opt] table args.root_keybindings The root (global) keybindings.
+-- @tparam[opt=false] boolean args.export_keybindings Create root (global) keybindings.
+-- @tparam[opt=false] boolean args.autostart Start the grabbing immediately
+-- @tparam[opt=false] boolean args.mask_modkeys Do not call the callbacks on
+--  modifier keys (like `Control` or `Mod4`) events.
+-- @function awful.keygrabber
+function keygrab.run_with_keybindings(args)
+    args = args or {}
+
+    local ret = gobject {
+        enable_properties   = true,
+        enable_auto_signals = true
+    }
+
+    rawset(ret, "_private", {})
+
+    -- Do not use `rawset` or `_private` because it uses the `gears.object`
+    -- auto signals.
+    ret.sequence = ""
+
+    gtable.crush(ret, keygrabber, true)
+
+    gtable.crush(ret, args)
+
+    ret._private.keybindings = {}
+    ret.release_event = args.release_event or "press"
+
+    -- Build the hook map
+    for _,v in ipairs(args.keybindings or {}) do
+        if #v >= 3 and #v <= 4 then
+            local _,key,callback = unpack(v)
+            if type(callback) == "function" then
+                ret._private.keybindings[key] = ret._private.keybindings[key] or {}
+                ret._private.keybindings[key][#ret._private.keybindings[key]+1] = v
+            else
+                gdebug.print_warning("The hook's 3rd parameter has to be a function.")
+            end
+        else
+            gdebug.print_warning("The hook has to have 4 parameters.")
```

Comment:
This warning does not really help finding the problem. How about `gears.debug.dump(v)` afterwards and changing the message accordingly?
Also, why "has to have 4 parameters", but above 3 is fine as well?